contracts/V4Router.sol

```diff
@@ -0,0 +1,186 @@
+                params.currencyIn = params.path[i].intermediateCurrency;
+            }
+
+            if (amountOut < params.amountOutMinimum) revert TooLittleReceived();
+        }
+    }
+
+    function _swapExactOutputSingle(IV4Router.ExactOutputSingleParams memory params, address msgSender) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountOut)),
+            params.sqrtPriceLimitX96,
+            msgSender,
+            true,
+            true,
+            params.hookData
+        );
+    }
+
+    function _swapExactOutput(IV4Router.ExactOutputParams memory params, address msgSender) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountIn;
+
+            for (uint256 i = pathLength; i > 0; i--) {
+                (PoolKey memory poolKey, bool oneForZero) =
+                    _getPoolAndSwapDirection(params.path[i - 1], params.currencyOut);
+                amountIn = uint128(
+                    -_swap(
+                        poolKey,
+                        !oneForZero,
+                        int256(int128(params.amountOut)),
+                        0,
+                        msgSender,
+                        i == 1,
+                        i == pathLength,
+                        params.path[i - 1].hookData
+                    )
+                );
+
+                params.amountOut = amountIn;
+                params.currencyOut = params.path[i - 1].intermediateCurrency;
+            }
+            if (amountIn > params.amountInMaximum) revert TooMuchRequested();
+        }
+    }
+
+    function _swap(
+        PoolKey memory poolKey,
+        bool zeroForOne,
+        int256 amountSpecified,
+        uint160 sqrtPriceLimitX96,
+        address msgSender,
+        bool settle,
+        bool take,
+        bytes memory hookData
+    ) private returns (int128 reciprocalAmount) {
+        BalanceDelta delta = poolManager.swap(
+            poolKey,
+            IPoolManager.SwapParams(
+                zeroForOne,
+                amountSpecified,
+                sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_PRICE + 1 : TickMath.MAX_SQRT_PRICE - 1)
+                    : sqrtPriceLimitX96
+            ),
+            hookData
+        );
+
+        if (zeroForOne) {
+            reciprocalAmount = amountSpecified < 0 ? delta.amount1() : delta.amount0();
+            if (settle) _payAndSettle(poolKey.currency0, msgSender, delta.amount0());
+            if (take) poolManager.take(poolKey.currency1, msgSender, uint128(delta.amount1()));
+        } else {
+            reciprocalAmount = amountSpecified < 0 ? delta.amount0() : delta.amount1();
+            if (settle) _payAndSettle(poolKey.currency1, msgSender, delta.amount1());
+            if (take) poolManager.take(poolKey.currency0, msgSender, uint128(delta.amount0()));
```

Comment:
so thoughts on this `msgSender` parameter
- what if we want to set a different recipient? sometimes in UR we want the recipient to be the UR itself, say if the next trade is then a V2 trade so we want the output of a V4 trade to be the UR so it can be used as an input to a new protocol. So i think we need a recipient parameter?
e.g. for v3 we do this
```
v3SwapExactInput(map(recipient), amountIn, amountOutMin, path, payer);
```
and that `map` function then figures out the recipient's address and passes it in. I think we need that here too.

- what if we want to set a different payer? again in UR sometimes the payer of the transaction should be the UR itself if this is in the second half of a trade that started on another protocol version. In that code snippet above, the `payer` parameter is the address of who is paying, which UR defines as
```
address payer = payerIsUser ? lockedBy : address(this);
```
I think we again need that possibility here? So I think we need to pass in `payer` and `recipient` instead of `msgSender`?